File: nim_core/ios/Classes/FLTService.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import Foundation

enum ServiceType: String {
  case MessageService
  case LifeCycleService = "InitializeService"
  case UserService
  case EventSubscribeService
  case ConversationService
  case SystemNotificationService = "SystemMessageService"
  case AudioRecordService = "AudioRecorderService"
  case AuthService
  case SessionService
  case TeamService
  case ChatroomService
  case NOSService
  case ChatExtendService
  case PassThroughService
  case SettingService = "SettingsService"
  case SuperTeamService
  case AvSignallingService
  case QChatObserver
  case QChatService
  case QChatServerService
  case QChatChannelService
  case QChatMessageService
  case QChatRoleService
  case QChatPushService
}

protocol FLTService {
  func serviceName() -> String

  func onMethodCalled(_ method: String, _ arguments: [String: Any],
                      _ resultCallback: ResultCallback)

  func register(_ nimCore: NimCore)

  func onInitialized()
}

File: nim_core_windows/windows/src/common/FLTService.cpp
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

#include "FLTService.h"

#include "FLTConvert.h"

std::string FLTService::getServiceName() const { return m_serviceName; }

void FLTService::notifyEvent(const std::string& eventName,
                             flutter::EncodableMap& arguments) {
  arguments.insert(std::make_pair(flutter::EncodableValue("serviceName"),
                                  flutter::EncodableValue(m_serviceName)));
  std::string strLog;
  strLog.append("en: ")
      .append(eventName)
      .append(", args: ")
      .append(Convert::getInstance()->getStringFormMapForLog(&arguments));
  std::list<std::string> logList;
  Convert::getInstance()->getLogList(strLog, logList);
  for (auto& it : logList) {
    YXLOG_API(Info) << it << YXLOGEnd;
  }
  NimCore::getInstance()->invokeMethod(eventName, arguments);
  YXLOG_API(Info) << "notifyEvent invoke completation." << YXLOGEnd;
}

void FLTService::notifyEventEx(const std::string& serviceName,
                               const std::string& eventName,
                               flutter::EncodableMap& arguments) {
  arguments.insert(std::make_pair(flutter::EncodableValue("serviceName"),
                                  flutter::EncodableValue(serviceName)));
  std::string strLog;
  strLog.append("en: ")
      .append(eventName)
      .append(", args: ")
      .append(Convert::getInstance()->getStringFormMapForLog(&arguments));
  std::list<std::string> logList;
  Convert::getInstance()->getLogList(strLog, logList);
  for (auto& it : logList) {
    YXLOG_API(Info) << it << YXLOGEnd;
  }
  NimCore::getInstance()->invokeMethod(eventName, arguments);
  YXLOG_API(Info) << "notifyEventEx invoke completation." << YXLOGEnd;
}
